gwcs/wcs.py

```diff
@@ -737,80 +748,53 @@ def to_fits_sip(self, bounding_box, max_error=None, max_rms=None, degree=None,
-        # transform2 = ((Shift(-crpix1) & Shift(-crpix2)) | Mapping([0,1,0,1]) |
-        #               (fit_poly_x & fit_poly_y) | Pix2Sky_TAN() |
-        #               RotateNative2Celestial(crval1, crval2, 180))
         return hdr
 
 
-def fit_2D_poly(degree, max_error, max_rms, u, v, x, y, verbose=False):
+def _fit_2D_poly(degree, max_error, max_rms, u, v, x, y, verbose=False):
     """
     Fit a pair of ordinary 2D polynomial to the supplied inputs (u, v) and
     outputs (x, y)
     """
     llsqfitter = LinearLSQFitter()
     # The case of one pass with the specified polynomial degree
-    if degree is not None:
-        poly_x = Polynomial2D(degree=degree)
-        poly_y = Polynomial2D(degree=degree)
+    if degree is None:
+        deglist = range(2,10)
+    else:
+        deglist = [degree]
+    for deg in deglist:
+        poly_x = Polynomial2D(degree=deg)
+        poly_y = Polynomial2D(degree=deg)
         fit_poly_x = llsqfitter(poly_x, u, v, x)
         fit_poly_y = llsqfitter(poly_y, u, v, y)
-        max_resid, rms = compute_distance_residual(x, y,
-                                                   fit_poly_x(u, v), fit_poly_y(u, v))
-    else:
-        cond = True
-        degree = 2
-        while cond:
-            poly_x = Polynomial2D(degree=degree)
-            poly_y = Polynomial2D(degree=degree)
-            fit_poly_x = llsqfitter(poly_x, u, v, x)
-            fit_poly_y = llsqfitter(poly_y, u, v, y)
-            max_resid, rms = compute_distance_residual(x, y,
-                                                       fit_poly_x(u, v), fit_poly_y(u, v))
+        max_resid, rms = _compute_distance_residual(x, y,
+                                                    fit_poly_x(u, v), fit_poly_y(u, v))
+        if verbose:
+            print(f'Degree = {degree}, max_resid = {max_resid}, rms = {rms}')
+        if ((max_error is not None and max_resid < max_error and max_rms is None)
+             or (max_error is not None and max_resid < max_error and max_rms is not None
+                 and rms < max_rms)
+             or (max_error is None and max_rms is not None and rms > max_rms)):
             if verbose:
-                print(f'Degree = {degree}, max_resid = {max_resid}, rms = {rms}')
-            if ((max_error is not None and max_resid < max_error and max_rms is None)
-                 or (max_error is not None and max_resid < max_error and max_rms is not None
-                     and rms < max_rms)
-                 or (max_error is None and max_rms is not None and rms > max_rms)):
-                cond = False
-                if verbose:
-                    print('terminating condition met')
-                break
-            else:
-                degree += 1
-                if degree == 10:
-                    raise RuntimeError("Fit conditions not met")
+                print('terminating condition met')
+            break
+        if deg == 9:
```

Comment:
You really must dislike `for-else` construction. I feel it is the most appropriate for this logic. Your alternative implementation would raise a `RuntimeError` **even if** `deg == 9` _provides a good fit_.

Resistance to `for-else` is futile.

Reply:
For your alternative implementation to work it would need to somehow know that this is the last execution of the loop and increment `deg` to 10 before exiting the loop normally so that the next `if deg == 10: raise RuntimeError()` would not trigger an exception for a good polynomial fit of 9th degree.

Reply:
Embrace Pythonic `for-else` 😄

Reply:
Notice that if it is a good fit for deg == 9, the break is executed before the test for deg == 9.

I've never been a big fan of the tentary construction, unless it is far clearer than all alternatives.

Reply:
Ah, I looked at this fragmented view instead of looking at the entire file/code.

One nitpick: instead of `deg == 9` maybe use `deg == deglist[-1]`.

Reply:
I think this logic is still flawed. What will happen if user asks to compute a fit with a polynomial of 9th order? Looking at the simplified/reduced code:
```python
def _fit_2D_poly(degree, max_error, u, v, x, y, verbose=False):
    # ...............
    if degree is None:
        deglist = range(1,10)
    else:
        deglist = [degree]
    for deg in deglist:
        # ..............
        max_resid = _compute_distance_residual(x, y, fit_poly_x(u, v), fit_poly_y(u, v))
        if max_error is not None and max_resid < max_error:
            break
        if deg == 9:
            raise RuntimeError("Unable...")
```
it seems to me that calling this function as `_fit_2D_poly(degree=9, max_error=None)` will result in a `RuntimeError`, isn't it? Wouldn't `deg == 9` BUT the `break` not happen before this last if due to `maxerr=None` hence `max_error is not None` would evaluate to `False` hence not reaching the `break`?

Reply:
Right you are. Fixed.